compiler/il/OMRNode.cpp

```diff
@@ -2357,7 +2357,115 @@ OMR::Node::isNotCollected()
    return false;
    }
 
+/**
+ * \brief
+ *    Check if the node is an internal pointer.
+ *
+ * \return
+ *    Return true if the node is an internal pointer, false otherwise.
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   if (!self()->getOpCode().hasPinningArrayPointer())
+      return false;
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collectable.
+ *
+ * \return
+ *    Return true if the node is collectable, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from a collectable reference, i.e. anything that is an address to a collected object.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (curNode->getOpCode().isTreeTop())
+      return false;
 
+   while (curNode)
+      {
+      if (!curNode->getType().isAddress())
+         return false;
+
+      if (curNode->isInternalPointer())
+         return true;
+
+      TR::ILOpCode op = curNode->getOpCode();
+      TR::ILOpCodes opValue = curNode->getOpCodeValue();
+      TR::Node *childNode = curNode->getFirstChild();
+
+      // If a language can handle a collected reference going via a non-address type,
+      // then the logic would need to be augmented to handle that
+      if (op.isConversion())
+         return false;
+
+      if (op.hasSymbolReference())
+         {
+         TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+         if (symbol->isInternalPointer() ||
+             symbol->isCollectedReference())
+             return true;
+         // Symbols for vft will have this flag set
+         if (symbol->isNotCollected())
+            return false;
+         }
+
+      // The following are all opcodes that are address type, non-TreeTop and non-conversion
+      switch (opValue)
+         {
+         case TR::aloadi:
+         case TR::aiadd:
+         case TR::aiuadd:
+         case TR::aladd:
+         case TR::aluadd:
+            curNode = curNode->getFirstChild();
+            break;
```

Comment:
Hopefully it's not actually valid for an address-type load to have a symbol with `!isInternalPointer() && !isCollectedReference() && !isNotCollected()`. If so, then `aloadi` (and `aload` and `aRegLoad`) nodes won't even reach here